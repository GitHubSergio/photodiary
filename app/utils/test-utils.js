/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import rootReducer from '../redux/reducers';
import { render } from '@testing-library/react-native';
// import { render } from 'react-native-testing-library';
import reduxStoreStateMocked from './redux-state';

export const renderWithReduxMockedStore = (ui, updatedStore) => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({ ...reduxStoreStateMocked, ...updatedStore });
  // console.log(store.getState());
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
  };
};

export const renderWithRedux = (
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
  } = {},
) => ({
  ...render(<Provider store={store}>{ui}</Provider>),
  store,
});
