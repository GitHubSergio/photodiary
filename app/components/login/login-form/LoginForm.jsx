// Imports
import React from 'react';
import { ScrollView, KeyboardAvoidingView } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useRoute } from '@react-navigation/native';

// Hooks
import { useInput } from '../../../utils/customHooks';

// Actions
import { signIn, signUp } from '../../../redux/middlewares/signingMiddleware';

// Components
import Logo from '../../logo/Logo';
import InputField from '../../shared/input-field/InputField';
import Button from '../../shared/buttons/Button';
import Loader from '../../shared/loader/Loader';
import Errors from '../../shared/errors/Errors';
import SwitchForm from './SwitchForm';

const LoginForm = () => {
  const { isSigning } = useSelector((state) => state.user);
  const { error } = useSelector((state) => state.user);
  const { name } = useRoute();

  const displayName = useInput('');
  const email = useInput('');
  const password = useInput('');
  const confirmPassword = useInput('');
  const dispatch = useDispatch();
  const isSignUpScreen = name === 'SignUpScreen';

  const handleSubmit = () => {
    isSignUpScreen
      ? dispatch(
          signUp(
            displayName.value,
            email.value,
            password.value,
            confirmPassword.value,
          ),
        )
      : dispatch(signIn(email.value, password.value));
  };

  return (
    <ScrollView>
      <KeyboardAvoidingView
        behavior="position"
        enabled
        keyboardVerticalOffset={70}>
        <Logo />

        {isSignUpScreen && (
          <InputField
            accessibilityLabel="username"
            placeHolder="Username"
            inputValue={displayName}
            onSubmitEditing={handleSubmit}
          />
        )}

        <InputField
          accessibilityLabel="email"
          placeHolder="Email"
          inputValue={email}
          onSubmitEditing={handleSubmit}
          keyboardType={'email-address'}
        />
        <InputField
          accessibilityLabel="password"
          placeHolder="Password"
          inputValue={password}
          onSubmitEditing={handleSubmit}
          secureTextEntry
        />

        {isSignUpScreen && (
          <InputField
            accessibilityLabel="confirm-password"
            placeHolder="Confirm Password"
            inputValue={confirmPassword}
            onSubmitEditing={handleSubmit}
            secureTextEntry
          />
        )}

        {isSigning ? (
          <Loader />
        ) : (
          <>
            <Button label="SUBMIT" onPress={handleSubmit} />
            <SwitchForm route={name} />
          </>
        )}
        {!!error && <Errors errorMsg={error} />}
      </KeyboardAvoidingView>
    </ScrollView>
  );
};

export default LoginForm;
