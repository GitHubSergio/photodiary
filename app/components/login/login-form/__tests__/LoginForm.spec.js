/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { renderWithRedux } from 'test-utils';
import { useNavigation, useRoute } from '@react-navigation/native';
import { fireEvent } from 'react-native-testing-library';
import LoginForm from '../LoginForm';

jest.mock('@react-navigation/native', () => ({
  useRoute: jest.fn(() => ({
    name: 'SignInScreen',
  })),
  useNavigation: jest.fn(),
}));

jest.mock('@react-native-firebase/auth', () => ({}));
jest.mock('@react-native-firebase/firestore', () => ({}));

jest.mock('react-native-image-picker', () => ({
  showImagePicker: jest.fn(),
}));

describe('<LoginForm />', () => {
  describe('@render', () => {
    describe('@render Log In', () => {
      it('should render the LoginScreen component', () => {
        const { toJSON } = renderWithRedux(<LoginForm />);
        expect(toJSON()).toMatchSnapshot();
      });

      it('should render the Loader component when attempting to login', () => {
        const {
          getByAccessibilityLabel,
          getByPlaceholder,
          toJSON,
        } = renderWithRedux(<LoginForm />);

        fireEvent.changeText(getByPlaceholder('Email'), 'test@mail.com');
        fireEvent.changeText(getByPlaceholder('Password'), '123456');
        fireEvent.press(getByAccessibilityLabel('submit-button'));
        expect(toJSON()).toMatchSnapshot();
      });

      it('should render the error component if email is missing', () => {
        const {
          getByAccessibilityLabel,
          getByPlaceholder,
          toJSON,
        } = renderWithRedux(<LoginForm />);

        fireEvent.changeText(getByPlaceholder('Password'), '123456');
        fireEvent.press(getByAccessibilityLabel('submit-button'));
        expect(toJSON()).toMatchSnapshot();
      });
    });

    describe('@render Sign Up', () => {
      it('should render the SignUp component', () => {
        useRoute.mockImplementation(
          jest.fn(() => ({
            name: 'SignUpScreen',
          })),
        );

        const { toJSON } = renderWithRedux(<LoginForm />);

        expect(toJSON()).toMatchSnapshot();
      });

      it('should render the Loader component when attempting to signup', () => {
        useRoute.mockImplementation(
          jest.fn(() => ({
            name: 'SignUpScreen',
          })),
        );
        useRoute.mockImplementation(
          jest.fn(() => ({
            name: 'SignUpScreen',
          })),
        );
        const {
          getByAccessibilityLabel,
          getByPlaceholder,
          toJSON,
        } = renderWithRedux(<LoginForm />);

        fireEvent.changeText(getByPlaceholder('Username'), 'MyUsername');
        fireEvent.changeText(getByPlaceholder('Email'), 'test@mail.com');
        fireEvent.changeText(getByPlaceholder('Password'), '123456');
        fireEvent.changeText(getByPlaceholder('Confirm Password'), '123456');
        fireEvent.press(getByAccessibilityLabel('submit-button'));
        expect(toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('@event', () => {
    fit('should navigate to the Sign Up screen when tapping the Sign Up button', () => {
      const navigation = useNavigation.mockImplementationOnce(() => ({
        navigate: jest.fn(),
      }));
      console.log(navigation.mock)
      useRoute.mockImplementation(
        jest.fn(() => ({
          name: 'LoginScreen',
        })),
      );
      const { getByTestId } = renderWithRedux(<LoginForm />);
      fireEvent.press(getByTestId('switch-btn'));
      expect(navigation).toHaveBeenCalled();
    });
    it('should navigate to the Sign In screen when tapping the Sign In button', () => {
      const navigation = useNavigation.mockImplementationOnce(() => ({
        navigate: jest.fn(),
      }));
      useRoute.mockImplementation(
        jest.fn(() => ({
          name: 'SignUpScreen',
        })),
      );
      const { getByTestId } = renderWithRedux(<LoginForm />);
      fireEvent.press(getByTestId('switch-btn'));
      expect(navigation).toHaveBeenCalled();
    });
  });
});
