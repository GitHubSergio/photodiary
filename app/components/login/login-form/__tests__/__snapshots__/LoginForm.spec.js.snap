// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> @render @render Log In should render the Loader component when attempting to login 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollView>
      <View>
        <View
          style={
            Object {
              "bottom": 0,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/logo-resized.png",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#2D2A4A",
                "borderRadius": 100,
                "marginBottom": 10,
              }
            }
          />
          <TextInput
            accessibilityLabel="email"
            allowFontScaling={true}
            keyboardType="email-address"
            placeholder="Email"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="test@mail.com"
          />
          <TextInput
            accessibilityLabel="password"
            allowFontScaling={true}
            placeholder="Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="123456"
          />
          <View
            testID="loader"
          >
            <ActivityIndicator
              animating={true}
              color="#0000ff"
              hidesWhenStopped={true}
              size="large"
            />
          </View>
        </View>
      </View>
    </ScrollView>
  </View>
</View>
`;

exports[`<LoginForm /> @render @render Log In should render the LoginScreen component 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollView>
      <View>
        <View
          style={
            Object {
              "bottom": 0,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/logo-resized.png",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#2D2A4A",
                "borderRadius": 100,
                "marginBottom": 10,
              }
            }
          />
          <TextInput
            accessibilityLabel="email"
            allowFontScaling={true}
            keyboardType="email-address"
            placeholder="Email"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            accessibilityLabel="password"
            allowFontScaling={true}
            placeholder="Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#2D2A4A",
                  "borderRadius": 10,
                  "height": 40,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessibilityLabel="submit-button"
              style={
                Array [
                  Object {
                    "color": "#F4F5F8",
                    "fontSize": 20,
                    "fontWeight": "bold",
                  },
                  undefined,
                ]
              }
            >
              SUBMIT
            </Text>
          </>
          <Button
            accessibilityLabel="switch-btn"
            title="Log In"
          />
        </View>
      </View>
    </ScrollView>
  </View>
</View>
`;

exports[`<LoginForm /> @render @render Log In should render the error component if email is missing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollView>
      <View>
        <View
          style={
            Object {
              "bottom": 0,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/logo-resized.png",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#2D2A4A",
                "borderRadius": 100,
                "marginBottom": 10,
              }
            }
          />
          <TextInput
            accessibilityLabel="email"
            allowFontScaling={true}
            keyboardType="email-address"
            placeholder="Email"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            accessibilityLabel="password"
            allowFontScaling={true}
            placeholder="Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="123456"
          />
          <
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#2D2A4A",
                  "borderRadius": 10,
                  "height": 40,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessibilityLabel="submit-button"
              style={
                Array [
                  Object {
                    "color": "#F4F5F8",
                    "fontSize": 20,
                    "fontWeight": "bold",
                  },
                  undefined,
                ]
              }
            >
              SUBMIT
            </Text>
          </>
          <Button
            accessibilityLabel="switch-btn"
            title="Log In"
          />
          <View
            style={
              Object {
                "padding": 10,
              }
            }
          >
            <Text
              accessibilityLabel="error-msg"
              style={
                Object {
                  "color": "#A51931",
                }
              }
            >
              Email is required
            </Text>
          </View>
        </View>
      </View>
    </ScrollView>
  </View>
</View>
`;

exports[`<LoginForm /> @render @render Sign Up should render the Loader component when attempting to signup 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollView>
      <View>
        <View
          style={
            Object {
              "bottom": 0,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/logo-resized.png",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#2D2A4A",
                "borderRadius": 100,
                "marginBottom": 10,
              }
            }
          />
          <TextInput
            accessibilityLabel="username"
            allowFontScaling={true}
            placeholder="Username"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="MyUsername"
          />
          <TextInput
            accessibilityLabel="email"
            allowFontScaling={true}
            keyboardType="email-address"
            placeholder="Email"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="test@mail.com"
          />
          <TextInput
            accessibilityLabel="password"
            allowFontScaling={true}
            placeholder="Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="123456"
          />
          <TextInput
            accessibilityLabel="confirm-password"
            allowFontScaling={true}
            placeholder="Confirm Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="123456"
          />
          <View
            testID="loader"
          >
            <ActivityIndicator
              animating={true}
              color="#0000ff"
              hidesWhenStopped={true}
              size="large"
            />
          </View>
        </View>
      </View>
    </ScrollView>
  </View>
</View>
`;

exports[`<LoginForm /> @render @render Sign Up should render the SignUp component 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollView>
      <View>
        <View
          style={
            Object {
              "bottom": 0,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../app/assets/images/logo-resized.png",
              }
            }
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#2D2A4A",
                "borderRadius": 100,
                "marginBottom": 10,
              }
            }
          />
          <TextInput
            accessibilityLabel="username"
            allowFontScaling={true}
            placeholder="Username"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            accessibilityLabel="email"
            allowFontScaling={true}
            keyboardType="email-address"
            placeholder="Email"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            accessibilityLabel="password"
            allowFontScaling={true}
            placeholder="Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <TextInput
            accessibilityLabel="confirm-password"
            allowFontScaling={true}
            placeholder="Confirm Password"
            placeholderTextColor="grey"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "marginBottom": 10,
                  "padding": 5,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#2D2A4A",
                  "borderRadius": 10,
                  "height": 40,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessibilityLabel="submit-button"
              style={
                Array [
                  Object {
                    "color": "#F4F5F8",
                    "fontSize": 20,
                    "fontWeight": "bold",
                  },
                  undefined,
                ]
              }
            >
              SUBMIT
            </Text>
          </>
          <Button
            accessibilityLabel="switch-btn"
            title="Log In"
          />
        </View>
      </View>
    </ScrollView>
  </View>
</View>
`;
